File: test/src/command/test/lint.c
/***********************************************************************************************************************************
Code Linter
***********************************************************************************************************************************/
#include <build.h>

#include <string.h>

#include "build/common/regExp.h"
#include "build/common/render.h"
#include "command/test/lint.h"
#include "common/log.h"
#include "storage/posix/storage.h"
#include "storage/storage.h"

/***********************************************************************************************************************************
Validate StringId values
***********************************************************************************************************************************/
static unsigned int
lintStrId(const String *const source)
{
    FUNCTION_LOG_BEGIN(logLevelDebug);
        FUNCTION_LOG_PARAM(STRING, source);
    FUNCTION_LOG_END();

    unsigned int result = 0;

    MEM_CONTEXT_TEMP_BEGIN()
    {
        // Check all StringIds
        RegExp *const strIdExp = regExpNew(STRDEF("STRID(5|6)(S){0,1}\\([^)]+\\)"));
        const char *sourcePtr = strZ(source);

        while (regExpMatch(strIdExp, STRDEF(sourcePtr)))
        {
            const String *const match = regExpMatchStr(strIdExp, STRDEF(sourcePtr));
            const StringList *const paramList = strLstNewSplitZ(
                strLstGet(strLstNewSplitZ(strSubN(match, 0, strSize(match) - 1), "("), 1), ",");
            const String *const param = strTrim(strLstGet(paramList, 0));

            sourcePtr = regExpMatchPtr(strIdExp, STRDEF(sourcePtr)) + strSize(match);

            // Skip macro definitions
            if (strEqZ(match, "STRID5(str, strId)") || strEqZ(match, "STRID5S(str, seq, strId)") ||
                strEqZ(match, "STRID6(str, strId)") || strEqZ(match, "STRID6S(str, seq, strId)"))
                continue;

            // Skip test strings
            if (strBeginsWithZ(param, "\\\"") && strEndsWithZ(param, "\\\""))
                continue;

            // Skip test values
            if (strLstSize(paramList) > 1 && strBeginsWithZ(strTrim(strLstGet(paramList, 1)), "TEST_"))
                continue;

            // Param must begin and end with a quote
            if (strZ(param)[0] != '"' || strZ(param)[strSize(param) - 1] != '"')
            {
                LOG_WARN_FMT("'%s' must have quotes around string parameter '%s'", strZ(match), strZ(param));
                result++;

                continue;
            }

            // Check validity of the string
            const String *expected;

            if (strBeginsWithZ(match, "STRID5" "(") || strBeginsWithZ(match, "STRID6" "("))
                expected = bldStrId(strZ(strSubN(param, 1, strSize(param) - 2)));
            else
            {
                expected = bldStrIdSeq(
                    strZ(strSubN(param, 1, strSize(param) - 2)), cvtZToUInt(strZ(strTrim(strLstGet(paramList, 1)))));
            }

            if (!strEq(match, expected))
            {
                LOG_WARN_FMT("'%s' should be '%s'", strZ(match), strZ(expected));
                result++;
            }
        }
    }
    MEM_CONTEXT_TEMP_END();

    FUNCTION_LOG_RETURN(UINT, result);
}

/***********************************************************************************************************************************
Detect characters that could be used to hide code, e.g. invisible, zero-width, bidirectional, or homoglyph characters. Source must
be 7-bit ASCII with tab and newline as the only permitted control characters. Intentional non-ASCII byte values must be written as
\xNN escapes so they remain visible in the source.
***********************************************************************************************************************************/
static unsigned int
lintChar(const String *const source)
{
    FUNCTION_LOG_BEGIN(logLevelDebug);
        FUNCTION_LOG_PARAM(STRING, source);
    FUNCTION_LOG_END();

    unsigned int result = 0;

    const unsigned char *const data = (const unsigned char *)strZ(source);
    const size_t size = strSize(source);
    unsigned int line = 1;

    for (size_t idx = 0; idx < size; idx++)
    {
        const unsigned char chr = data[idx];

        // Count lines for reporting
        if (chr == '\n')
        {
            line++;
            continue;
        }

        // Tab is the only other permitted control character
        if (chr == '\t')
            continue;

        // Flag any byte outside printable 7-bit ASCII. Invisible, bidirectional, and homoglyph characters all live here, so they
        // are rejected wholesale and intentional non-ASCII must be written with \xNN escapes.
        if (chr < 0x20 || chr > 0x7E)
        {
            // Decode the code point for reporting on a best-effort basis. Validation is unnecessary since the character is rejected
            // regardless, so the byte count is taken from the lead byte and limited to the bytes available.
            unsigned int seqSize = 1;

            if (chr >= 0xF0)
                seqSize = 4;
            else if (chr >= 0xE0)
                seqSize = 3;
            else if (chr >= 0xC0)
                seqSize = 2;

            if (seqSize > size - idx)
                seqSize = (unsigned int)(size - idx);

            unsigned int codePoint = chr;

            if (seqSize > 1)
            {
                codePoint = chr & (0xFF >> (seqSize + 1));

                for (unsigned int seqIdx = 1; seqIdx < seqSize; seqIdx++)
                    codePoint = (codePoint << 6) | (data[idx + seqIdx] & 0x3F);
            }

            LOG_WARN_FMT("line %u contains disallowed character U+%04X (source must be 7-bit ASCII)", line, codePoint);
            result++;

            idx += seqSize - 1;
        }
    }

    FUNCTION_LOG_RETURN(UINT, result);
}

/***********************************************************************************************************************************
Files that are exempt from the content checks. Binary files are unscannable places to hide content, so each entry must be a
deliberate, reviewable decision.
***********************************************************************************************************************************/
static const char *const lintFileSkipList[] =
{
    "doc/resource/logo.png",                                        // Project logo
    "test/data/filecopy.table.bin",                                 // Binary test fixture
    "doc/resource/git-history.cache",                               // Generated from git history, contains non-ASCII author names
};

static bool
lintFileSkip(const String *const name)
{
    FUNCTION_TEST_BEGIN();
        FUNCTION_TEST_PARAM(STRING, name);
    FUNCTION_TEST_END();

    bool result = false;

    for (unsigned int skipIdx = 0; skipIdx < LENGTH_OF(lintFileSkipList); skipIdx++)
    {
        if (strEqZ(name, lintFileSkipList[skipIdx]))
        {
            result = true;
            break;
        }
    }

    FUNCTION_TEST_RETURN(BOOL, result);
}

/**********************************************************************************************************************************/
void
lintAll(const String *const pathRepo)
{
    FUNCTION_LOG_BEGIN(logLevelDebug);
        FUNCTION_LOG_PARAM(STRING, pathRepo);
    FUNCTION_LOG_END();

    ASSERT(pathRepo != NULL);

    MEM_CONTEXT_TEMP_BEGIN()
    {
        const Storage *const storageRepo = storagePosixNewP(pathRepo);
        StorageIterator *const repoItr = storageNewItrP(
            storageRepo, NULL, .level = storageInfoLevelType, .recurse = true, .sortOrder = sortOrderAsc);

        while (storageItrMore(repoItr))
        {
            const StorageInfo info = storageItrNext(repoItr);

            // Skip anything that is not a file and files that are exempt from the content checks
            if (info.type != storageTypeFile || lintFileSkip(info.name))
                continue;

            // Read the file, skipping any that cannot be read. Committed files are always readable after checkout, so this only
            // occurs from local filesystem state, and readability is enforced by the build rather than the linter.
            const Buffer *content = NULL;

            TRY_BEGIN()
            {
                content = storageGetP(storageNewReadP(storageRepo, info.name));
            }
            CATCH(FileOpenError)
            {
                // File cannot be read, so there is nothing to check
            }
            TRY_END();

            if (content == NULL)
                continue;

            unsigned int errorTotal = 0;

            // A binary file cannot be scanned for hidden content, so it is not allowed unless added to lintFileSkipList
            if (memchr(bufPtrConst(content), '\0', bufUsed(content)) != NULL)
            {
                LOG_WARN("unexpected binary file (add to the linter skip list if intentional)");
                errorTotal = 1;
            }
            // Otherwise the file must be 7-bit ASCII text, with the StringId check applied to C source
            else
            {
                const String *const source = strNewBuf(content);

                errorTotal = lintChar(source);

                // The StringId check applies to C source, including hand-written .c.inc, but not generated or vendored includes
                if (strEndsWithZ(info.name, ".c") || strEndsWithZ(info.name, ".h") ||
                    (strEndsWithZ(info.name, ".c.inc") && !strEndsWithZ(info.name, ".auto.c.inc") &&
                     !strEndsWithZ(info.name, ".vendor.c.inc")))
                {
                    errorTotal += lintStrId(source);
                }
            }

            if (errorTotal > 0)
                THROW_FMT(FormatError, "%u linter error(s) in '%s' (see warnings above)", errorTotal, strZ(info.name));
        }
    }
    MEM_CONTEXT_TEMP_END();

    FUNCTION_LOG_RETURN_VOID();
}

File: test/lib/command/lint/lint.py
"""Code Linter.

Scans every file in the repository for content that could hide code from review, checks that no line runs past the project line
length, checks that StringId macros encode what they claim to, checks that a block macro is closed by the macro that matches it, and
checks that every test module is declared in define.yaml.

Everything found is reported as a warning and the number of them is returned rather than raised, so the run continues. A linter
reads the source as it is written, which means a mistake in one place shows up wherever the source stopped making sense to it: a
misspelled block macro is skipped rather than opening a block, so the next closer is reported instead, hundreds of lines from the
line that needs fixing. The compiler places the same mistake exactly, so let the run reach it. Whatever runs the linter is
responsible for failing once it has."""

####################################################################################################################################
import os
import re

from command.lint.ascii import lint_ascii
from command.lint.line import lint_line
from command.lint.macro import lint_macro
from command.lint.string_id import lint_str_id
from command.test.define import TEST_MODULE_PATH, test_def_file, test_def_parse
from common.log import *
from common.storage import path_list_recurse

# Files that are exempt from the content checks. A binary file is an unscannable place to hide content, so each entry must be a
# deliberate, reviewable decision.
_FILE_SKIP_LIST = (
    "doc/resource/card.png",  # Card shown where a link to the documentation is posted
    "test/data/filecopy.table.bin",  # Binary test fixture
    "doc/resource/git-history.cache",  # Generated from git history, contains non-ASCII author names
)

# Paths that are exempt from the line length check, matched against the start of the name so an entry that ends in a slash exempts
# everything under it. The check is excludes rather than includes so a file that is added is checked without anyone remembering to
# add it, which is the way lines got long in the first place.
_LINE_SKIP_LIST = (
    "doc/manifest.xml",  # Documentation variables, which are a value per line
    "doc/resource/",  # Images and caches rather than source
    "doc/xml/",  # Documentation, which the coding standards exempt
    "test/certificate/",  # Generated certificates and keys
    "test/data/",  # Test fixtures
    "test/uncrustify.cfg",  # Written by uncrustify rather than by hand
)

# Extensions that are exempt from the line length check
_LINE_SKIP_EXT = (
    ".md",  # Markdown, written a paragraph per line so the editor wraps it and a diff is per paragraph
    ".vendor.h",  # Vendored source, which is kept as it came so it can be updated from upstream
    ".vendor.c.inc",
)

# A python module in a tool library, e.g. build/lib/common/render.py
_LIB_MODULE_EXP = re.compile(r"^([^/]+)/lib/(.+\.py)$")


####################################################################################################################################
def _lint_line_apply(name):
    """Should the line length check be applied to this file?"""

    return not name.startswith(_LINE_SKIP_LIST) and not name.endswith(_LINE_SKIP_EXT)


####################################################################################################################################
def _lint_c_apply(name):
    """Should the checks that read C source be applied to this file?

    They apply to C source, including hand-written .c.inc, but not to generated or vendored includes. Generated source cannot drift
    from what generated it and vendored source is kept as it came, so neither is ours to fix."""

    if name.endswith(".c") or name.endswith(".h"):
        return True

    return name.endswith(".c.inc") and not name.endswith(".auto.c.inc") and not name.endswith(".vendor.c.inc")


####################################################################################################################################
def cmd_lint(path_repo):
    """Lint every file in the repository and return the number of errors found."""

    result = 0
    lib_module = {}

    # File each test module declared in define.yaml lives in. A test module that is not declared is never built or run, and nothing
    # else reports it since the file is simply never read.
    test_module = {test_def_file(module) for module in test_def_parse(path_repo)}

    for name in path_list_recurse(path_repo):
        # Everything where the test modules live is a test module, so it must be declared in define.yaml
        if name.startswith(TEST_MODULE_PATH) and name not in test_module:
            log(WARN, "test module '%s' is not defined in test/define.yaml" % name)
            result += 1

        # A module may appear in only one library. Python resolves a module name to the first library on the path that has it and
        # ignores the rest, so a duplicate would hide the shadowed module from every tool with no indication that it had.
        match = _LIB_MODULE_EXP.match(name)

        if match is not None:
            lib, module = match.groups()

            if module in lib_module:
                log(WARN, "module '%s' is in the %s and %s libraries" % (module, lib_module[module], lib))
                result += 1

            lib_module[module] = lib

        # Skip files that are exempt from the content checks
        if name in _FILE_SKIP_LIST:
            continue

        # Read the file, skipping any that cannot be read. Committed files are always readable after checkout, so this only occurs
        # from local filesystem state, and readability is enforced by the build rather than the linter.
        try:
            with open(os.path.join(path_repo, name), "rb") as file:
                data = file.read()
        except OSError:
            continue

        # A binary file cannot be scanned for hidden content, so it is not allowed unless added to the skip list
        if b"\x00" in data:
            log(WARN, "unexpected binary file (add to the linter skip list if intentional)")
            error_total = 1
        # Otherwise the file must be 7-bit ASCII text, with the line length check applied to everything that is not exempt and the
        # StringId and block macro checks applied to C source
        else:
            error_ascii = lint_ascii(data)
            error_total = error_ascii
            source = data.decode("utf-8", errors="replace")

            if _lint_line_apply(name):
                error_total += lint_line(source)

            if _lint_c_apply(name):
                error_total += lint_str_id(source)

                # The block macro check reads every character in the file rather than scanning for a pattern, so it is applied only
                # to source that is ASCII. What it would report about a character it cannot read is what the ASCII check above has
                # already reported, in worse words.
                if error_ascii == 0:
                    error_total += lint_macro(source)

        # Name the file the warnings above are for, since they report the line they are on and nothing else
        if error_total > 0:
            log(WARN, "%u linter error(s) in '%s'" % (error_total, name))
            result += error_total

    return result
